lib/testLauncher/repl.js

```diff
@@ -0,0 +1,174 @@
+const net = require('net');
+const path = require('path');
+const repl = require('repl');
+const colors = require('colors');
+const chokidar = require('chokidar');
+
+const {getShortFunctionBody} = require('../utils/stringUtils');
+const SuitestError = require('../utils/SuitestError');
+const {REPL_START, REPL_STOP, REPL_INIT} = require('../constants/repl');
+const texts = require('../texts');
+
+const exceptionListeners = {};
+const errorEvents = ['uncaughtException', 'unhandledRejection'];
+
+function pauseErrorListeners() {
+	// return;
+	errorEvents.forEach(event => {
+		if (!exceptionListeners[event])
+			exceptionListeners[event] = process.rawListeners(event).map(one => one);
+
+		process.removeAllListeners(event);
+		process.on(event, err => console.error(err));
+	});
+}
+
+function restoreErrorListeners() {
+	errorEvents.forEach(event => {
+		if (exceptionListeners[event]) {
+			process.removeAllListeners(event);
+			exceptionListeners[event].forEach(func => process.on(event, func));
+		}
+	});
+}
+
+/**
+ * Establishes communication channel with the main process and
+ * starts the REPL
+ * @param {Object} opts - repl opts see below
+ * @returns {Promise<any>} - resolves when repl session ends
+ */
+function startRepl(opts) {
+	const {
+		replIpcErrorInChildProcess,
+		replIpcNotAvailable,
+		replIpcNotImplemented,
+	} = texts;
+
+	if (!process.env.REPL_IPC_PORT) {
+		throw new SuitestError(
+			replIpcNotAvailable(),
+			SuitestError.IPC_ERROR
+		);
+	}
+
+	const ipcSocket = net.connect(process.env.REPL_IPC_PORT, 'localhost').unref();
+
```

Comment:
We set up the ipcSocket  to make it work with spawn ?

Reply:
We need a channel between the main process and the child in order for the main process to know when it should switch the terminal to the raw mode and back. Mode switching is necessary so that the control commands go directly to the child process and NOT go to the main process.

When we were using fork node was managing this for us, however using fork with shared io does not work as expected and with piped io it kind of did although the terminal behaved weird from time to time. Plus the inconvenience with not being able to use the nonjs command launchers such as "mocha".

IPC communication looks like the following:
- Main process sets up TCP server
- Main process passes the server port number to the child process via env variable.
- Child process connects (when suitest.interactive starts) and submits a handshake
- Main process responds with the isTTY value (since it is not available for reading from the child process)
- Child process sets up REPL terminal and sends message again
- Main process switches terminal to raw mode.
- On repl exit child process sends another notification and main process repipes io and switches raw mode off.

As a bonus if ever the API should be ported to another language the communication can stay the same.